File: backend/src/db/pool.js
require('dotenv').config({ override: true });
const { Pool, types } = require('pg');
const logger = require('../utils/logger');

// pg parses SQL DATE columns (OID 1082) into a JS Date at local midnight,
// which then serializes to UTC and shifts across a day boundary for any
// timezone ahead of UTC (e.g. reminder_date '2026-08-03' becomes
// '2026-08-02T18:30:00.000Z' in IST) — return the raw 'YYYY-MM-DD' string
// instead so date-only columns round-trip exactly.
types.setTypeParser(1082, (value) => value);

// Most managed Postgres providers (Render, Heroku, RDS with enforced SSL)
// require SSL and reject plain connections outright — DB_SSL lets a
// production deploy opt in without code changes. rejectUnauthorized is left
// configurable since many providers issue certs that don't chain to a
// standard root store the way it's set up by default.
const useSsl = process.env.DB_SSL === 'true';

const pool = new Pool({
  host:     process.env.DB_HOST     || 'localhost',
  port:     parseInt(process.env.DB_PORT || '5432'),
  database: process.env.DB_NAME     || 'fieldtrack',
  user:     process.env.DB_USER     || 'postgres',
  password: process.env.DB_PASSWORD || '',
  ssl:      useSsl ? { rejectUnauthorized: process.env.DB_SSL_REJECT_UNAUTHORIZED !== 'false' } : false,
  max:                  parseInt(process.env.DB_POOL_MAX || '10'),
  idleTimeoutMillis:    parseInt(process.env.DB_IDLE_TIMEOUT_MS || '30000'),
  connectionTimeoutMillis: parseInt(process.env.DB_CONNECTION_TIMEOUT_MS || '5000'),
});

pool.on('error', (err) => {
  logger.error('Unexpected PostgreSQL client error', { error: err.message, stack: err.stack });
});

module.exports = pool;

File: backend-py/app/db/pool.py
"""
pool.py — asyncpg connection pool, raw SQL only (no ORM), mirroring
src/db/pool.js exactly (same env vars, same defaults, same SSL handling).

Note (NOTES.md): pool.js installs a custom type parser for SQL DATE columns
(OID 1082) to work around `pg` returning a JS Date at local midnight, which
then serializes to UTC and drifts across a day boundary in IST. asyncpg has
no equivalent bug — it returns Python `datetime.date` objects, which FastAPI/
Pydantic serialize as plain 'YYYY-MM-DD' strings with no timezone attached —
so no workaround is needed here. Call `.isoformat()` explicitly wherever a
route hand-builds a dict instead of returning a Pydantic model, to keep
response shapes identical to the Node backend's plain date strings.
"""
import ssl as ssl_lib

import asyncpg

from app.core import config
from app.core.logging_config import log_error

_pool: asyncpg.Pool | None = None


def _build_ssl_arg():
    """Mirrors pool.js's `ssl: useSsl ? { rejectUnauthorized: ... } : false`.

    BUG FIXED HERE (found live against Render's managed Postgres — never
    exercised by local testing, which always ran DB_SSL=false): asyncpg's
    `ssl` parameter takes `True` / `False` / an `ssl.SSLContext`, NOT the
    string `"require"`. The original version of this function passed
    `ssl="require"` (invalid) when DB_SSL_REJECT_UNAUTHORIZED was true, and
    — worse — passed `ssl=True` (full certificate verification) when
    DB_SSL_REJECT_UNAUTHORIZED was FALSE, which is backwards: `True` makes
    asyncpg verify the server's certificate against the system trust store,
    which fails immediately for Render's managed Postgres cert (it doesn't
    chain to a standard root CA — the exact reason Node's own
    `rejectUnauthorized: false` exists). That failed handshake surfaces as
    `asyncpg.exceptions.ConnectionDoesNotExistError: connection was closed
    in the middle of operation` — not an obviously SSL-shaped error message,
    which is why this took a real deploy to surface.
    """
    if not config.DB_SSL:
        return None
    if config.DB_SSL_REJECT_UNAUTHORIZED:
        return True
    ctx = ssl_lib.create_default_context()
    ctx.check_hostname = False
    ctx.verify_mode = ssl_lib.CERT_NONE
    return ctx


async def connect() -> None:
    global _pool
    _pool = await asyncpg.create_pool(
        host=config.DB_HOST,
        port=config.DB_PORT,
        database=config.DB_NAME,
        user=config.DB_USER,
        password=config.DB_PASSWORD,
        ssl=_build_ssl_arg(),
        min_size=1,
        max_size=config.DB_POOL_MAX,
        max_inactive_connection_lifetime=config.DB_IDLE_TIMEOUT_MS / 1000,
        timeout=config.DB_CONNECTION_TIMEOUT_MS / 1000,
    )


async def disconnect() -> None:
    global _pool
    if _pool is not None:
        await _pool.close()
        _pool = None


def get_pool() -> asyncpg.Pool:
    if _pool is None:
        raise RuntimeError("DB pool accessed before connect() — app not started correctly.")
    return _pool


async def fetch(query: str, *args):
    try:
        return await get_pool().fetch(query, *args)
    except Exception as err:  # noqa: BLE001 — mirrors pool.on('error') best-effort logging
        log_error("Unexpected PostgreSQL client error", error=str(err))
        raise


async def fetchrow(query: str, *args):
    return await get_pool().fetchrow(query, *args)


async def execute(query: str, *args) -> str:
    return await get_pool().execute(query, *args)
